model/waterfall.go

```diff
@@ -13,54 +13,111 @@ import (
-	Activated   bool            `bson:"activated" json:"activated,omitempty"`
-	DisplayName string          `bson:"display_name" json:"display_name,omitempty"`
-	Version     string          `bson:"version" json:"version,omitempty"`
-	Tasks       []WaterfallTask `bson:"tasks" json:"tasks,omitempty"`
+	Activated   bool            `bson:"activated" json:"activated"`
+	DisplayName string          `bson:"display_name" json:"display_name"`
+	Version     string          `bson:"version" json:"version"`
+	Tasks       []WaterfallTask `bson:"tasks" json:"tasks"`
 }
 
 type WaterfallBuildVariant struct {
 	Id          string           `bson:"_id" json:"_id"`
-	DisplayName string           `bson:"display_name" json:"display_name,omitempty"`
-	Builds      []WaterfallBuild `bson:"builds" json:"builds,omitempty"`
+	DisplayName string           `bson:"display_name" json:"display_name"`
+	Builds      []WaterfallBuild `bson:"builds" json:"builds"`
 }
 
 type WaterfallOptions struct {
 	Limit      int      `bson:"-" json:"-"`
+	MaxOrder   int      `bson:"-" json:"-"`
+	MinOrder   int      `bson:"-" json:"-"`
 	Requesters []string `bson:"-" json:"-"`
 }
 
-// GetWaterfallVersions returns at most `opts.limit` versions for a given project.
-// TODO: It will eventually return `opts.limit` activated versions that also satisfy any given filters.
-func GetWaterfallVersions(ctx context.Context, projectId string, opts WaterfallOptions) ([]Version, error) {
+// GetActiveWaterfallVersions returns at most `opts.limit` activated versions for a given project.
+func GetActiveWaterfallVersions(ctx context.Context, projectId string, opts WaterfallOptions) ([]Version, error) {
 	invalidRequesters, _ := utility.StringSliceSymmetricDifference(opts.Requesters, evergreen.SystemVersionRequesterTypes)
 	if len(invalidRequesters) > 0 {
 		return nil, errors.Errorf("invalid requester(s) '%s'; only commit-level requesters can be applied to the waterfall query", invalidRequesters)
 	}
+
+	if opts.MaxOrder != 0 && opts.MinOrder != 0 {
+		return nil, errors.New("cannot provide both max and min order options")
+	}
+
 	match := bson.M{
 		VersionIdentifierKey: projectId,
 		VersionRequesterKey: bson.M{
 			"$in": opts.Requesters,
 		},
+		VersionActivatedKey: true,
+	}
+
+	pagingForward := opts.MaxOrder != 0
+	pagingBackward := opts.MinOrder != 0
+
+	if pagingForward {
+		match[VersionRevisionOrderNumberKey] = bson.M{"$lt": opts.MaxOrder}
+	} else if pagingBackward {
+		match[VersionRevisionOrderNumberKey] = bson.M{"$gt": opts.MinOrder}
+	}
+
+	pipeline := []bson.M{{"$match": match}}
+
+	if pagingBackward {
+		pipeline = append(pipeline, bson.M{"$sort": bson.M{VersionRevisionOrderNumberKey: 1}})
+		pipeline = append(pipeline, bson.M{"$limit": DefaultWaterfallQueryCount})
+		pipeline = append(pipeline, bson.M{"$sort": bson.M{VersionRevisionOrderNumberKey: -1}})
+
+	} else {
+		pipeline = append(pipeline, bson.M{"$sort": bson.M{VersionRevisionOrderNumberKey: -1}})
```

Comment:
I might be getting mixed up by the sorting/pagination dir ection, but isn't this sorted backwards rather than forwards? Is it intentionally paging forward but sorted backwards?

Or is this meant like "paging" like in the sense of the Spruce page, where you click forward (i.e. left to right) but you're actually going backward through the waterfall history?

Reply:
The latter! We return versions sorted in descending order (newest to oldest) for the waterfall